src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -23,7 +24,7 @@ public SolutionStateChecksums(params object[] children) : base(WellKnownSynchron
         public Checksum Info => (Checksum)Children[0];
         public ProjectChecksumCollection Projects => (ProjectChecksumCollection)Children[1];
 
-        public void Find(
+        public async ValueTask FindAsync(
```

Comment:
❕ Prefer `async Task` to `async ValueTask` unless you are _specifically_ leveraging the caching feature of `ValueTask` with a custom task source. The async state machine will optimize `async Task` to return `Task.CompletedTask` without allocations when returning synchronously.

```suggestion
        public async Task FindAsync(
```

Reply:
This is intentionally using `ValueTask` since @heejaechang is concerned about Task allocations here.

Reply:
this particular one is mostly okay since it is Task without any return value, but ones that call these Find methods need to be ValueTask<T> so that they don't create task unnecessarily.

Reply:
@tmat `async Task` does not allocate on any path where `async ValueTask` is allocation-free.

Reply:
📝 The problem with `ValueTask` is it places special requirements on the caller which are not obvious. For example:

1. The result _must_ be awaited for correctness (fire-and-forget can leak)
2. The result must _only_ be awaited once (multiple awaits can produce incorrect behavior)

Reply:
> The async state machine will optimize async Task to return Task.CompletedTask without allocations when returning synchronously.

I am not sure how deep that optimization is. is await machinery actually check whether calling await return synchronously or is it just check await code path being called?

so, only reason this code path needs to be async is GetTextAsync and we do cache text once it is read, so except the very first call, it is synchronous even if signature is GetTextAsync. best is making GetTextAsync ValueTask and make all caller ValueTask to the top which will remove bunch of Task allocations.

Reply:
I see, the ValueTask optimization is not needed to void returning methods since Task is already optimized for that case.

Reply:
if you guys care that much, probably time to debug in. I did once and mostly one that matter is one that return value.

> 1. The result must be awaited for correctness (fire-and-forget can leak)
> 2. The result must only be awaited once (multiple awaits can produce incorrect behavior)

I am not sure what is leaking and 2 seems weird. is it talking about holding valuetask and calling await twice on it? looking at the code - https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Threading/Tasks/ValueTask.cs#L259 - it should work fine?

Reply:
by looking our implementation of GetTextAsync, it will return Task.From(...) that mark returned task to be completed, and await machinary will make caller to be synchronous code path from there using Task.CompletedTask cache for void one. and we cache Task.From(...) itself for text to reduce Task allocation.

so, it will be allocation free up to GetRemoteDataAsync. that will need to return ValueTask.

Reply:
Yes, the constraints on ValueTask are explained [here](https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/).

It might be useful to have an analyzer that disallows any use of method returning ValueTask other than awaiting it.

Reply:
You mean something like dotnet/roslyn-analyzers#3001? 😄